File: android/core/src/test/kotlin/com/ridelink/core/testutil/Vectors.kt
package com.ridelink.core.testutil

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import java.io.File

/**
 * Loads shared golden vectors from `protocol/vectors/` at the repo root. The Kotlin tests must
 * execute these files directly rather than duplicating vector data as Kotlin literals
 * (CLAUDE.md "Shared protocol vectors — not optional").
 */
object Vectors {
    private val vectorsDir: File by lazy {
        val path =
            requireNotNull(System.getProperty("ridelink.protocolVectorsDir")) {
                "ridelink.protocolVectorsDir system property not set — see core/build.gradle.kts"
            }
        File(path)
    }

    fun load(relativePath: String): JsonElement {
        val file = File(vectorsDir, relativePath)
        require(file.exists()) { "vector file not found: ${file.absolutePath}" }
        return Json.parseToJsonElement(file.readText())
    }
}

File: ios/Packages/RideLinkCore/Tests/RideLinkCoreTests/TestSupport/Vectors.swift
import Foundation

/// Loads shared golden vectors from `protocol/vectors/` at the repo root. The Swift tests must
/// execute these files directly rather than duplicating vector data as Swift literals
/// (CLAUDE.md "Shared protocol vectors — not optional").
enum Vectors {
    enum VectorsError: Error {
        case repoRootNotFound
    }

    /// Walks upward from this source file's own location (captured at compile time via
    /// `#filePath`) until it finds a directory containing `protocol/vectors` — robust against
    /// exactly where in the package tree the test file lives, with no build-tool wiring needed
    /// (unlike the Android side, which passes the path via a Gradle system property).
    private static func repoRoot(from callerFilePath: String = #filePath) throws -> URL {
        var dir = URL(fileURLWithPath: callerFilePath).deletingLastPathComponent()
        let fileManager = FileManager.default
        while true {
            let candidate = dir.appendingPathComponent("protocol/vectors")
            if fileManager.fileExists(atPath: candidate.path) {
                return dir
            }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path {
                throw VectorsError.repoRootNotFound
            }
            dir = parent
        }
    }

    static func loadJSON(_ relativePath: String) throws -> Any {
        let root = try repoRoot()
        let url = root.appendingPathComponent("protocol/vectors").appendingPathComponent(relativePath)
        let data = try Data(contentsOf: url)
        return try JSONSerialization.jsonObject(with: data, options: [.fragmentsAllowed])
    }
}
